daemon/user/daemon_test.go

```diff
@@ -0,0 +1,166 @@
+ *
+ */
+
+package user_test
+
+import (
+	"time"
+
+	"github.com/godbus/dbus"
+	"github.com/snapcore/snapd/daemon/user"
+
+	"fmt"
+
+	. "gopkg.in/check.v1"
+)
+
+type daemonSuite struct {
+	i      int
+	args   [][]string
+	errors []error
+	outs   [][]byte
+}
+
+var _ = Suite(&daemonSuite{})
+
+type exportData struct {
+	iface interface{}
+	name  string
+	path  dbus.ObjectPath
+}
+
+type mockDBusConn struct {
+	exportError      error
+	exportData       []*exportData
+	requestNameReply dbus.RequestNameReply
+	requestNameError error
+	requestNameData  struct {
+		name  string
+		flags dbus.RequestNameFlags
+	}
+	signalData struct {
+		signals []chan<- *dbus.Signal
+	}
+}
+
+func (c *mockDBusConn) Export(v interface{}, path dbus.ObjectPath, iface string) error {
+	c.exportData = append(c.exportData, &exportData{iface: v, path: path, name: iface})
+	return c.exportError
+}
+
+func (c *mockDBusConn) RequestName(name string, flags dbus.RequestNameFlags) (dbus.RequestNameReply, error) {
+	c.requestNameData.name = name
+	c.requestNameData.flags = flags
+	return c.requestNameReply, c.requestNameError
+}
+
+func (c *mockDBusConn) Signal(ch chan<- *dbus.Signal) {
+	c.signalData.signals = append(c.signalData.signals, ch)
+}
+
+func (c *mockDBusConn) sendTerminateSignal() {
+	for _, ch := range c.signalData.signals {
+		close(ch)
+	}
+}
+
+func (c *mockDBusConn) Close() error {
+	return nil
+}
+
+func (s *daemonSuite) TestStartsAndRequestsName(c *C) {
+	conn := &mockDBusConn{
+		requestNameReply: dbus.RequestNameReplyPrimaryOwner,
+	}
+
+	user.MockSessionBus(conn, nil)
+
+	d, err := user.NewDaemon()
+	c.Assert(err, IsNil)
+	c.Assert(d, NotNil)
+
+	d.Start()
+	time.Sleep(time.Second * 1)
+	conn.sendTerminateSignal()
+	d.Stop()
+
+	c.Assert(conn.requestNameData.name, Equals, "com.canonical.SafeLauncher")
+	c.Assert(conn.requestNameData.flags, Equals, dbus.NameFlagDoNotQueue)
+
+	c.Assert(len(conn.exportData), Equals, 2)
+	c.Assert(conn.exportData[0].name, Equals, "com.canonical.SafeLauncher")
+	c.Assert(conn.exportData[0].path, Equals, dbus.ObjectPath("/"))
+	c.Assert(conn.exportData[0].iface, NotNil)
+	c.Assert(conn.exportData[1].name, Equals, "org.freedesktop.DBus.Introspectable")
+	c.Assert(conn.exportData[1].path, Equals, dbus.ObjectPath("/"))
+	c.Assert(conn.exportData[1].iface, NotNil)
+}
+
+func (s *daemonSuite) TestStartupFailsWhenSessionBusIsNotAvailable(c *C) {
+	user.MockSessionBus(nil, fmt.Errorf("Session bus is not available"))
+
+	d, err := user.NewDaemon()
+	c.Assert(err, IsNil)
+	c.Assert(d, NotNil)
+
+	d.Start()
+	time.Sleep(time.Second * 1)
+	err = d.Stop()
+	c.Assert(err, ErrorMatches, "Session bus is not available")
+}
+
+func (s *daemonSuite) TestStartupFailsWhenNameRequestFails(c *C) {
+	conn := &mockDBusConn{
+		requestNameError: fmt.Errorf("Failed to request name"),
+	}
+
+	user.MockSessionBus(conn, nil)
+
+	d, err := user.NewDaemon()
+	c.Assert(err, IsNil)
+	c.Assert(d, NotNil)
+
+	d.Start()
+	time.Sleep(time.Second * 1)
```

Comment:
I think we can do better than sleeping. Perhaps add a notification channel/callback to notify when we're ready?